### PR TITLE
add service worker

### PR DESCRIPTION
**Pull request by TomWoodward**
for: openstax/unified#1035

### CODE REVIEW

File: src/helpers/applicationUpdates.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    window.location.reload(true);
+  }
+};
+
+export const activateSwAndReload = (sw: ServiceWorkerRegistration | undefined) => () => {
+  const {waiting} = sw || {waiting:  undefined};
+
+  if (waiting) {
+    waiting.addEventListener('statechange', function activatingHandler() {
+      if (waiting.state === 'activating') {
```
Comment (janKozie1) [inline src/helpers/applicationUpdates.ts:64]: What's the reason its on `activating` and not `activated`?

File: src/app/notifications/components/UpdatesAvailable.tsx [hunk 2]
```diff
@@ -1,29 +1,48 @@
...
-const reload = () => {
-  if (typeof(window) !== 'undefined') {
-    window.location.reload(true);
+/*
+ * the service worker will update on its own after a refresh, but if we're going to
+ * prompt people we want to make sure its ready before the refresh
+ */
+
+// tslint:disable-next-line:variable-name
+const UpdatesAvailable = ({className}: {className ?: string}) => {
```
Comment (janKozie1) [inline src/app/notifications/components/UpdatesAvailable.tsx:18]: Unnecessary space before `?:`

**Review by janKozie1 — APPROVED** [review]
Just minor questions

File: src/helpers/applicationUpdates.ts [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+  // if there is already a waiting or installing service worker, don't try to udpate it.
+  // it'll still be activated by activateSwAndReload
+  if (sw.waiting || sw.installing) {
+    return false;
+  }
+
+  return true;
+};
+
+export const findAndInstallServiceWorkerUpdate = (sw: ServiceWorkerRegistration | undefined, callback: () => void) => {
```
[the remark this answers is not in the document]
Comment (TomWoodward) [reply src/helpers/applicationUpdates.ts:20]: it depends on how you define "serviceWorker", the `navigator.serviceWorker` variable is actually the [ServiceWorkerContainer](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerContainer), which handles migrating between multiple service worker scripts. the registration represents an actual service worker script running

File: src/helpers/applicationUpdates.ts [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    })
+    .catch((e) => {
+      Sentry.captureException(e);
+      callback();
+    })
+  ;
+};
+
+export const forceReload = () => {
+  if (typeof(window) !== 'undefined') {
```
[the remark this answers is not in the document]
Comment (TomWoodward) [reply src/helpers/applicationUpdates.ts:54]: the typeof makes it not throw an undefined variable exception if its not defined, doing `if (window)` will throw in nodejs

File: src/helpers/applicationUpdates.ts (hunk 1 above)
Comment (TomWoodward) [reply src/helpers/applicationUpdates.ts:64]: the [active](https://developer.mozilla.org/en-US/docs/Web/API/ServiceWorkerRegistration/active) slot catchings things that are both `activating` and `active`

**Comment by TomWoodward** [comment]
@janKozie1 i had to change some stuff to get CI to pass, you might want to review again

**Pull request merged**